p2p: No delay in adding fixed seeds if -dnsseed=0 and peers.dat is empty

dhruv opened the pull request:
Closes #19795 

Before PR: If `peers.dat` is empty and `-dnsseed=0`, bitcoind will fallback on to fixed seeds but only after a 60 seconds delay.
After PR: There's no 60 second delay.

To reproduce:
`rm ~/.bitcoin/peers.dat && src/bitcoind -dnsseed=0` without and with patch code

Other changes in the PR:
- `-fixedseeds` command line argument added: `-dnsseed=0 -fixedseeds=0 -addnode=X` provides a trusted peer only setup. `-dnsseed=0 -fixedseeds=0` allows for a `addnode` RPC to add a trusted peer without falling back to hardcoded seeds.

dhruv committed:
[p2p] No delay in adding fixed seeds if -dnsseed=0 and peers.dat is empty. Add -fixedseeds arg.

n-thumann reviewed (approved):
tACK bcbd637 on master at 3ba25e3, was able to reproduce the issue and confirm that it´s fixed with this PR ✌️

dhruv commented:
@practicalswift I'm sorry this took so long. Ready for review.

dhruv commented:
@n-thumann Thank you, for the prompt review.

practicalswift commented on src/net.cpp:
```diff
@@ -1841,19 +1841,22 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+        // Here we allow 60 seconds for the DNS seeds to return results. If they do not
+        // return any results, or if they were disabled with -dnsseed=0, we add
+        // the fixed seeds.
+        bool dnsseed = gArgs.GetBoolArg("-dnsseed", DEFAULT_DNSSEED);
```
Nit:

```suggestion
        const bool dnsseed = gArgs.GetBoolArg("-dnsseed", DEFAULT_DNSSEED);
```

dhruv replied:
Done. Thanks for the catch.

practicalswift commented on src/net.cpp:
```diff
@@ -1841,19 +1841,22 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+        bool dnsseed = gArgs.GetBoolArg("-dnsseed", DEFAULT_DNSSEED);
+        if (addrman.size() == 0 &&
+            ((GetTime() - nStart > 60) || !dnsseed)) {
+            if (!dnsseed) {
```
Nit: Personally I find `if-else` logic easier to read if starting with the truthy case. In this specific case: starting with the `if (dnsseed) {` case and handle the `!dnsseed` case in the `else` block :)

dhruv replied:
Done

practicalswift commented:
ACK bcbd637bcfe45d94fbe27688a35d071e6bcd8411

Feel free to ignore my nits :)

Thanks for fixing this issue!

jonatack commented on src/net.cpp:
```diff
@@ -1841,19 +1841,22 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+        // the fixed seeds.
+        bool dnsseed = gArgs.GetBoolArg("-dnsseed", DEFAULT_DNSSEED);
+        if (addrman.size() == 0 &&
+            ((GetTime() - nStart > 60) || !dnsseed)) {
```
Might be outside the scope of this PR, but ISTM`GetTime()` is deprecated; see `src/util/time.h`.

jonatack commented on src/net.cpp:
```diff
@@ -1841,19 +1841,22 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+        // return any results, or if they were disabled with -dnsseed=0, we add
+        // the fixed seeds.
+        bool dnsseed = gArgs.GetBoolArg("-dnsseed", DEFAULT_DNSSEED);
+        if (addrman.size() == 0 &&
```
```suggestion
        if (addrman.empty() &&
```

jonatack replied:
Oops, you are right. Probably out of scope to create it unless you want to.

jonatack reviewed (commented):
Concept ACK. Maybe add a regression test in `feature_config_args.py` to cover these cases.

practicalswift commented:
@dhruv I saw you removed the `done` logic: are you sure that `!addrman.empty()` is _guaranteed_ after `addrman.Add(convertSeed6(Params().FixedSeeds()), local);`? AFAICT that condition must be hold for the new code to be equivalent to the old code with.

dhruv reviewed (commented):
@jonatack Added a regression test in feature_config_args.py to cover the cases. It's a slow test though because we have to wait 60 seconds. Is there anything I can do to reduce that interval for tests? I thought about making it configurable via cli flags but I will look to your guidance on whether that is standard practice.

dhruv commented:
> @dhruv I saw you removed the `done` logic: are you sure that `!addrman.empty()` is _guaranteed_ after `addrman.Add(convertSeed6(Params().FixedSeeds()), local);`? AFAICT that condition must be hold for the new code to be equivalent to the old code.

@practicalswift Thank you for catching that. I overlooked the implication of the `static` qualifier on `done` and mistakenly thought the code was redundant. It is now added back in.

It seems that the `done` logic is used to make sure that `addrman.Add(convertSeed6(Params().FixedSeeds()), local);` gets called no more than once regardless of the outcome. When I followed the code in `CAddrMan`, it seems that the invocation just adds the fixed seeds to the address tables so long as the seeds are well-formed. So invoking the deterministic function more than once is unnecessary. I'm new around here, so I'm hoping you can confirm my understanding.

dhruv commented:
Appveyor has [passed](https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/35050140) on `e7de5b5`, Github is just not reflecting it.

jonatack reviewed (commented):
ACK e7de5b53e7b4433d41b519ca04d4c3aad0b6e186

Thanks for adding the tests. No strong opinion whether it's better to drop the slow first test or keep it. There is a `setmocktime` RPC often used in the functional tests, and a `-mocktime=<n>` config option that could be passed an `extra_arg` to `start_node`, but AFAICT neither one will have any effect and I'm not sure it's worth changing the code to be more testable unless you have an elegant way to do it or it could be generalized and useful elsewhere.

A couple of nit comments follow; you can ignore them unless you need to retouch for other reasons.

ajtowns commented:
> If `peers.dat` is empty and `-dnsseed=0`

Is this a use case worth optimising for? The linked issue doesn't really provide a reason why this use case is worthwhile, as far as I can see, just that the filer thought it was surprising.

The only case I can see where you'd want to set `-dnsseed=0` is if you somehow don't have DNS despite being connected to the real internet, or if you're really worried about privacy. But if you're really worried about privacy, connecting to fixed seeds is probably worse than doing a dns query, and if you were doing something clever, like triggering an outbound/inbound connection to your own node, losing the 60 second delay might also drop privacy.

dhruv commented on src/net.cpp:
```diff
@@ -1826,10 +1826,10 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
 
     // Minimum time before next feeler connection (in microseconds).
-    int64_t nNextFeeler = PoissonNextSend(nStart*1000*1000, FEELER_INTERVAL);
+    int64_t nNextFeeler = PoissonNextSend(nStart.count()*1000*1000, FEELER_INTERVAL);
```
Ah, I did see that. Wasn't sure I should lint fix a line I didn't make any other changes on as it would change the git blame output. What is the standard practice?

dhruv commented:
> Is this a use case worth optimising for? The linked issue doesn't really provide a reason why this use case is worthwhile, as far as I can see, just that the filer thought it was surprising.

Fair question, @ajtowns. Let's page @practicalswift and see if he can shed light on his original use case.

practicalswift commented:
> The only case I can see where you'd want to set `-dnsseed=0` is if you somehow don't have DNS despite being connected to the real internet, or if you're really worried about privacy. But if you're really worried about privacy, connecting to fixed seeds is probably worse than doing a dns query, and if you were doing something clever, like triggering an outbound/inbound connection to your own node, losing the 60 second delay might also drop privacy.

There are multiple important uses cases for `-dnsseed=0`:
* You don't want to use clearnet at all (`-dnsseed=0 -onlynet=onion -listen=0`)
* You don't want to broadcast to the DNS seed operators (or anyone listening on the path to those DNS seed operators) that you're starting a new node (`-dnsseed=0 -listen=0`)
* You want to test how Bitcoin Core behaves if the DNS seeds are unavailable

I hope you're not advocating the removal of `-dnsseed=0` :)

> Is this a use case worth optimising for? The linked issue doesn't really provide a reason why this use case is worthwhile, as far as I can see, just that the filer thought it was surprising.

I'm not sure I would call removal of a 60 second sleep an optimization. To me it is clearly a bug fix :)

Also, please note that the behaviour is unchanged if `-dnsseed=0` is not set.

The only effect of this change is that users of `-dnsseed=0` won't have to wait an extra 60 seconds for their node to start.

Given the above clarification: do your skepticism to this bug fix remain? :)

sipa commented:
> You don't want to use clearnet at all (-dnsseed=0 -onlynet=onion -listen=0)

FWIW, if you use `-proxy`, DNS queries go over the proxy, including DNS seeding.

practicalswift commented:
> > You don't want to use clearnet at all (-dnsseed=0 -onlynet=onion -listen=0)
> 
> FWIW, if you use `-proxy`, DNS seeding happens over the proxy (which in the case of Tor means the exit node will do the resolving for you).

Yes I know, but then you'll have to trust the exit node operator not to tamper with the DNS responses :)

sipa commented:
Of course, and I'm not commenting on the validity of the use case in general. Just that "You don't want to use clearnet at all" on its own is not a reason to use -dnsseed=0.

practicalswift commented:
> Of course, and I'm not commenting on the validity of the use case in general. Just that "You don't want to use clearnet at all" on its own is not a reason to use -dnsseed=0.

Fair point. I should have written "You don't want to use clearnet at all and you don't trust exit node operators" :)

practicalswift commented:
> FWIW, if you use `-proxy`, DNS seeding happens over the proxy (which in the case of Tor means the exit node will do the resolving for you).

By "DNS seeding" you're in this context referring to the "establish `ADDR_FETCH` connection via proxy to seed nodes"-logic, right?

AFAICT we've never supported actual DNS queries to the DNS seed nodes via proxy.

practicalswift reviewed (approved):
Tested ACK e7de5b53e7b4433d41b519ca04d4c3aad0b6e186

This fixes the bug which makes `-dnsseed=0` users have to wait out a 60 second sleep intended only for `-dnsseed=1` users :)

sipa commented:
@practicalswift Yes, when -proxy is enabled, no actual DNS seeding is done, but ADDR_FETCH connections are established to the seeder's names instead. That was the reason for introducing ADDR_FETCH connections (formerly called oneshot connections) initially.

ajtowns commented:
> Fair point. I should have written "You don't want to use clearnet at all and you don't trust exit node operators" :)

Ah, and are thus only connecting to the ~89 fixed seeds with onion addresses, and not going via exit nodes at all? That makes more sense.

I think that currently, if you do `-onlynet=onion -listen=0 -dnsseed=0 -addnode=mytrustedpeer.onion`,  and you successfully connect to mytrustedpeer.onion within 60 seconds, you won't add the fixed seeds (your addrman will contain mypeer.onion rather than being empty), whereas with this change you will add the fixed seeds. Not sure that's a change for the better?

Perhaps rather than a time check, it would be better to check for something like "if addrman is empty, there are no oneshots left to try, no open connections and no added nodes to keep trying" ?

OTOH, if it is a change for the better, it might make sense to move the add fixed nodes code to its own function, and call it from ThreadDNSAddressSeed directly (after trying all the dns seeds, with addrman still empty, and only after the opencon thread as processed any oneshots that got queued up), and from CConnman::Start when  `-dnsseed=0` and addrman is empty. That way you avoid unnecessary waiting in all cases, I think.

dhruv commented:
My apologies for the delay in response. I needed to study all the ways seeding is done before I felt like I could have an opinion. I audited the code and found 6 ways `addrman` grows. I want to document them here for future readers. Reviewers well-versed in seeding can skip to the next section.

### Peer seeding sources
1. `-connect`: If the user runs the node in this configuration, `bitcoind` will only ever make outbound connections with nodes specified in `-connect`. It will still make `getaddr` RPC calls to them and `addrman` will grow, but no other outbound connections will be made. `-dnsseed` and `-listen` are disabled unless overridden by the user.
2. `peers.dat`: `addrman` flushes to `peers.dat`. When not in `-connect` mode, the node will use it's pre-existing knowledge of possible peers from `peers.dat`. Each peer will be queried using `getaddr` and `addrman` will grow further.
3. `-dnsseed`: defaults to true. Uses some domain names owned by Bitcoin contributors to expand the peers in `addrman`. Domains can be found in `chainparams.cpp:CMainParams::CMainParams`. Only `getaddr` calls are made to these nodes. Connections are of type ADDR_FETCH.
4. `-seednode`: Like `-dnsseed`, but uses user-specified peers instead of dns lookup. 
5. `-addnode`: Like `-connect`, except when `addrman` grows, new outbound peers will be established. (This won't happen if `-connect` is also specified).
6. Fixed seeds: If all else fails, there are hardcoded seeds in `chainparamsseeds.h` that the node can try to seed peers from.

### Continued PR discussion
- `-connect` is not relevant as seeding does not really matter in that case.
- We are solving for the case when `peers.dat` is empty
- We are solving for the case when `-dnsseed=0`
- That leaves us to deal with `-seednode` and `-addnode`

@ajtowns has made a great observation that in the case that `-seednode` or `-addnode` are provided, removing the 60 second delay before trying the fixed seeds makes it so that we are relying on a last-resort mechanism before giving the nodes in `-seednode` and `-addnode` a chance to help us grow `addrman`. I've updated the code to reflect this. We will now try to use fixed seeds when:

```
-connect is not provided

AND peers.dat is empty

AND ((-dnsseed=0 AND all_ADDR_FETCH_processed AND -addnode=[]) OR 60 seconds have passed)
```

This gives DNS seeds(ADDR_FETCH), -seednode connections(also ADDR_FETCH) and -addnode peers 60 seconds to help us find peers IF they are specified. Note that we explicitly need to check if `-dnsseed=0` even though it may seem like `all_ADDR_FETCH_processed` covers that case. This is because when `-dnsseed=1`, `ThreadDNSAddressSeed` will add ADDR_FETCH connections almost instantaneously, but `ThreadOpenConnections` may check this condition before the DNS seeds are added. 

@ajtowns Please note that I was not able to do this:
> OTOH, if it is a change for the better, it might make sense to move the add fixed nodes code to its own function, and call it from ThreadDNSAddressSeed directly (after trying all the dns seeds, with addrman still empty, and only after the opencon thread as processed any ADDR_FETCH nodes that got queued up), and from CConnman::Start when -dnsseed=0 and addrman is empty. That way you avoid unnecessary waiting in all cases, I think.

because I don't understand how we decide we are giving up on `-addnode` manual connections being able to grow `addrman` without a timeout (in this case, 60 seconds).

This PR is ready for re-review.

luke-jr commented:
`addnode` is also a RPC method. It is quite plausible that an external peer discovery mechanism might launch bitcoind and use it to get connections started within that 60 seconds...

ajtowns commented:
> >  it might make sense to move the add fixed nodes code to its own function, ...

I was thinking something like https://github.com/ajtowns/bitcoin/commits/202010-fixed-seeds 

> I don't understand how we decide we are giving up on -addnode manual connections being able to grow addrman without a timeout (in this case, 60 seconds).

The above makes adding fixed seeds always be an immediate decision -- either because it's the only way of connecting to the network at startup, or because you tried DNS seeds and they all failed. If you want to disable that because you're doing `-addnode` later via RPC, you could specify `-dnsseed=0 -seednode=127.0.0.1`, though perhaps adding a `-fixedseeds=0` param would be better.

I think the argument for removing the timeout is just that if you end up adding fixed seeds, the entire time before you did that is wasted because you weren't able to connect to the p2p network (if you _were_ able to, you wouldn't have ended up adding the fixed seeds), so rather than waste people's time, just make the decision earlier.

I think that's probably better behaviour for the edge cases too -- if you're doing `-addnode` to a trusted peer in order to avoid untrusted seeds, then you probably don't want to fall back to the untrusted peers if you happen to not be able to connect to your trusted peer in the first minute of trying.

ajtowns commented on src/net.cpp:
```diff
@@ -1826,10 +1826,10 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
 
     // Minimum time before next feeler connection (in microseconds).
-    int64_t nNextFeeler = PoissonNextSend(nStart*1000*1000, FEELER_INTERVAL);
+    int64_t nNextFeeler = PoissonNextSend(nStart.count() * 1000 * 1000, FEELER_INTERVAL);
```
Use `count_microseconds`

dhruv replied:
Done

ajtowns commented on src/net.cpp:
```diff
@@ -1841,18 +1841,35 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+            }
+
+            const bool dnsseed = gArgs.GetBoolArg("-dnsseed", DEFAULT_DNSSEED);
+            // Checking !dnsseed is cheaper before locking 2 mutexes.
```
`GetBoolArg` takes the `cs_args` lock, so probably better to move this check outside of the loop if you care about that

dhruv replied:
Done

ajtowns commented on src/net.cpp:
```diff
@@ -1841,18 +1841,35 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+            // Checking !dnsseed is cheaper before locking 2 mutexes.
+            if (!addFixedSeeds && !dnsseed) {
+                LOCK2(m_addr_fetches_mutex, cs_vAddedNodes);
+                if(!dnsseed && m_addr_fetches.empty() && vAddedNodes.empty()) {
```
`if (` -- needs a space

laanwj commented:
Concept ACK.

dhruv commented:
Thank you, for the review and for taking the time to teach me via your branch, @ajtowns.

I've added a new command line flag: `-fixedseeds`. A user can run `-dnsseed=0 -fixedseeds=0` and then issue a `addnode` RPC to satisfy the use case @luke-jr mentioned. `-dnsseed=0 -fixedseeds=0 -addnode=X` will provide the trusted peer only behavior @ajtowns mentioned. I also added a functional test. 

I definitely see the appeal of not having any wait, but I chose not to split up adding fixed seeds into `CConnman::Start` and `ThreadDNSAddressSeed` yet. If the user runs `-dnsseed=0 -seednode=X` or `-dnsseed=0 -addnode=X`, `CConnMan::Start` will not add fixed seeds, and `ThreadDNSAddressSeed` will not run at all. Then, in case the node at X is unreachable/censored, our node would never fall back to fixed seeds. This would be a departure from how things work today as fallback to untrusted peers is the default behavior. Perhaps we can allow for the trusted-peer-only behavior using `-dnsseed=0 -fixedseeds=0 -addnode=X`, but not impose it? In addition, `-connect` already provides that option. If other reviewers also agree that we should default to the trusted-peer-only behavior, I am open to changing it.

The changes to use chrono for feelers are included in #20044 and nicely bundled with other chrono types, so I did not include it.

Ready for further review.

dhruv commented:
Rebased

LarryRuane commented on src/net.cpp:
```diff
@@ -1895,18 +1903,34 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
-            if (!done) {
-                LogPrintf("Adding fixed seed nodes as DNS doesn't seem to be available.\n");
+        // Static bool makes sure we only add fixed seeds once in binary lifetime.
+        static bool fixedSeedsDone = disableFixedSeeds;
```
I like this renaming of the variable `done` to `fixedSeedsDone`, but maybe we can take the opportunity to remove this `static`? Static variables are scope-limited global variables (in general, globals should be avoided). 
<details>
<summary>suggestion</summary>

```diff
-    const bool disableFixedSeeds = !gArgs.GetBoolArg("-fixedseeds", DEFAULT_FIXEDSEEDS);
+    bool addFixedSeeds = gArgs.GetBoolArg("-fixedseeds", DEFAULT_FIXEDSEEDS);
 
-    if (disableFixedSeeds) {
+    if (!addFixedSeeds) {
         LogPrintf("Fixed seeds are disabled\n");
     }
 
     while (!interruptNet)
     {
         ProcessAddrFetch();
 
         if (!interruptNet.sleep_for(std::chrono::milliseconds(500)))
             return;
 
         CSemaphoreGrant grant(*semOutbound);
         if (interruptNet)
             return;
 
         // Static bool makes sure we only add fixed seeds once in binary lifetime.
-        static bool fixedSeedsDone = disableFixedSeeds;
-        if (!fixedSeedsDone && addrman.size() == 0) {
+        if (addFixedSeeds && addrman.size() == 0) {
             // When the node starts with an empty peers.dat, there are a few other sources of peers before
             // we fallback on to fixed seeds: -dnsseed, -seednode, -addnode
             // If none of those are available, we fallback on to fixed seeds immediately, else we allow
             // 60 seconds for any of those sources to populate addrman.
-            bool addFixedSeeds = false;
+            bool addFixedSeedsNow = false;
             // It is cheapest to check if enough time has passed first.
             if (GetTime<std::chrono::seconds>() > nStart + std::chrono::minutes{1}) {
-                addFixedSeeds = true;
+                addFixedSeedsNow = true;
                 LogPrintf("Adding fixed seeds as 60 seconds have passed and addrman is empty\n");
             }
 
             // Checking !dnsseed is cheaper before locking 2 mutexes.
-            if (!addFixedSeeds && !dnsseed) {
+            if (!addFixedSeedsNow && !dnsseed) {
                 LOCK2(m_addr_fetches_mutex, cs_vAddedNodes);
                 if (m_addr_fetches.empty() && vAddedNodes.empty()) {
-                    addFixedSeeds = true;
+                    addFixedSeedsNow = true;
                     LogPrintf("Adding fixed seeds as -dnsseed=0, -addnode is not provided and and all -seednode(s) attempted\n");
                 }
             }
 
-            if (addFixedSeeds) {
+            if (addFixedSeedsNow) {
                 CNetAddr local;
                 local.SetInternal("fixedseeds");
                 addrman.Add(convertSeed6(Params().FixedSeeds()), local);
-                fixedSeedsDone = true;
+                addFixedSeeds = false;
             }
         }
```
</details>

Should new variables be written in `snake_case`?

dhruv replied:
Done! I too prefer the elimination of the static variable. Wasn't sure reviewers would be keen on it, but I agree it is cleaner.

LarryRuane commented on test/functional/feature_config_args.py:
```diff
@@ -145,11 +146,61 @@ def test_networkactive(self):
+        # So after 60 seconds, the node should fallback to fixed seeds (this is a slow test)
+        assert not os.path.exists(os.path.join(default_data_dir, "peers.dat"))
+        start = time.time()
+        with self.nodes[0].assert_debug_log(expected_msgs=["Loaded 0 addresses from peers.dat",
+                "0 addresses found from DNS seeds", "Adding fixed seeds as 60 seconds have passed and addrman is empty"],
+                timeout=80):
```
```suggestion
        with self.nodes[0].assert_debug_log(expected_msgs=[
                "Loaded 0 addresses from peers.dat",
                "0 addresses found from DNS seeds",
                "Adding fixed seeds as 60 seconds have passed and addrman is empty",
            ],
                timeout=80):
```
(and similarly below)

dhruv replied:
Done.

LarryRuane commented on test/functional/feature_config_args.py:
```diff
@@ -145,11 +146,61 @@ def test_networkactive(self):
         self.stop_node(0)
 
+    def test_seed_peers(self):
+        self.log.info('Test seed peers')
```
```suggestion
        self.log.info('Test seed peers, this will take about 2 minutes')
```

LarryRuane commented on test/functional/feature_config_args.py:
```diff
@@ -145,11 +146,61 @@ def test_networkactive(self):
+        # We expect the node will fallback immediately to fixed seeds
+        assert not os.path.exists(os.path.join(default_data_dir, "peers.dat"))
+        start = time.time()
+        with self.nodes[0].assert_debug_log(expected_msgs=["Loaded 0 addresses from peers.dat",
+                "DNS seeding disabled",
+                "Adding fixed seeds as -dnsseed=0, -addnode is not provided and and all -seednode(s) attempted\n"]):
```
You can also check for the absence of debug.log messages (I'm not sure if it's worth it in this case):
```suggestion
        with self.nodes[0].assert_debug_log(expected_msgs=[
                "Loaded 0 addresses from peers.dat",
                "DNS seeding disabled",
                "Adding fixed seeds as -dnsseed=0, -addnode is not provided and and all -seednode(s) attempted\n"
            ], unexpected_msgs=["60 seconds"]):
```

dhruv replied:
Changed the indentation. I did not add the `unexpected_msgs=["60 seconds"]` as it is implicit in the 2 second timeout for `assert_debug_log`.

LarryRuane reviewed (commented):
As suggested in the description, I tested:
```
rm ~/.bitcoin/peers.dat && src/bitcoind -dnsseed=0
```
without the patch (note the timestamps):
```
2021-01-02T00:00:49Z msghand thread start
2021-01-02T00:01:50Z Adding fixed seed nodes as DNS doesn't seem to be available.
```
With the patch:
```
2021-01-02T00:05:38Z msghand thread start
2021-01-02T00:05:38Z Adding fixed seeds as -dnsseed=0, -addnode is not provided and and all -seednode(s) attempted
```
Ran `clang-format-diff.py` (it found nothing). Nice PR! Suggestions are minor and optional.
ACK eb1eef4e21fda019a9d309656aac1ffaf04b262d
(but take with a grain of salt; I don't know this area of the code well.)

dhruv commented on src/init.cpp:
```diff
@@ -435,8 +435,9 @@ void SetupServerArgs(NodeContext& node)
     argsman.AddArg("-discover", "Discover own IP addresses (default: 1 when listening and no -externalip or -proxy)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-dns", strprintf("Allow DNS lookups for -addnode, -seednode and -connect (default: %u)", DEFAULT_NAME_LOOKUP), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    argsman.AddArg("-dnsseed", "Query for peer addresses via DNS lookup, if low on addresses (default: 1 unless -connect used)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-dnsseed", strprintf("Query for peer addresses via DNS lookup, if low on addresses (default: %u unless -connect used)", DEFAULT_DNSSEED), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Done. Thanks!

dhruv commented on src/init.cpp:
```diff
@@ -435,8 +435,9 @@ void SetupServerArgs(NodeContext& node)
-    argsman.AddArg("-dnsseed", "Query for peer addresses via DNS lookup, if low on addresses (default: 1 unless -connect used)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-dnsseed", strprintf("Query for peer addresses via DNS lookup, if low on addresses (default: %u unless -connect used)", DEFAULT_DNSSEED), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-externalip=<ip>", "Specify your own public address", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-fixedseeds", strprintf("Allow fixed seeds if DNS seeds don't provide peers (default: %u)", DEFAULT_FIXEDSEEDS), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Done

dhruv commented:
Thanks for the review @LarryRuane, and my apologies for the long wait.

Comments addressed. Rebased. Ready for further review.

dhruv commented:
Fixed failing linter.

LarryRuane commented:
re-reviewed, re-tested
ACK 3fdea1f77b7afa07789139f17c0e03ee55f3ee02

laanwj commented:
Code review ACK 3fdea1f77b7afa07789139f17c0e03ee55f3ee02

laanwj commented on test/functional/feature_config_args.py:
```diff
@@ -147,11 +148,67 @@ def test_networkactive(self):
+                "DNS seeding disabled",
+                "Adding fixed seeds as -dnsseed=0, -addnode is not provided and and all -seednode(s) attempted\n"]):
+            self.start_node(0, extra_args=['-dnsseed=0'])
+        assert time.time() - start < 5
```
It's slightly risky to have a test that depends on real time instead of mocked times. CI environments can be extremely slow, so this bound might introduce random failures.

dhruv replied:
You're right - that could make for a flaky test. I picked 5 seconds as an arbitrary bound. What I _really_ wanted to check was that the node did not wait 60 seconds to use fixed seeds.

```suggestion
        assert time.time() - start < 60
```

accomplished that. And if CI is so slow that 60 seconds pass, the test will fail anyway because the logs will say "Adding fixed seeds as 60 seconds have passed and addrman is empty" instead of "Adding fixed seeds as -dnsseed=0, -addnode is not provided and and all -seednode(s) attempted".

laanwj replied:
Thanks, good idea!

LarryRuane commented:
re-ACK fe3e993968d6b46777d5a16a662cd22790ddf5bb

dhruv commented:
Updated tests for CI stability. Ready for further review.

laanwj commented:
re-ACK fe3e993968d6b46777d5a16a662cd22790ddf5bb
checked that the only change since my last ACK is the `<` time bounds for the test

MarcoFalke commented on test/functional/feature_config_args.py:
```diff
@@ -147,11 +148,67 @@ def test_networkactive(self):
+        with self.nodes[0].assert_debug_log(expected_msgs=[
+                "Loaded 0 addresses from peers.dat",
+                "DNS seeding disabled",
+                "Adding fixed seeds as -dnsseed=0, -addnode is not provided and and all -seednode(s) attempted\n"]):
```
`and and`

dhruv replied:
Please see #21165

MarcoFalke commented on test/functional/feature_config_args.py:
```diff
@@ -147,11 +148,67 @@ def test_networkactive(self):
+                "0 addresses found from DNS seeds",
+                "Adding fixed seeds as 60 seconds have passed and addrman is empty"], timeout=80):
+            self.start_node(0, extra_args=['-dnsseed=1'])
+        assert time.time() - start >= 60
```
wouldn't it work to use `setmocktime` inside the context manager above and check here: `node.uptime() >= 60` to speed up the test?

jonatack replied:
I looked at this last September (https://github.com/bitcoin/bitcoin/pull/19884#pullrequestreview-486061184) and it didn't look feasible, interesting if it is.

MarcoFalke replied:
```
$ time test/functional/feature_config_args.py 
2021-02-12T17:02:39.647000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_kuphxowf
2021-02-12T17:02:45.294000Z TestFramework (INFO): Test config args logging
2021-02-12T17:02:45.605000Z TestFramework (INFO): Test seed peers, this will take about 2 minutes
2021-02-12T17:02:48.587000Z TestFramework (INFO): Test -networkactive option
2021-02-12T17:02:54.023000Z TestFramework (INFO): Stopping nodes
2021-02-12T17:02:54.127000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_kuphxowf on exit
2021-02-12T17:02:54.127000Z TestFramework (INFO): Tests successful

real	0m14.682s
user	0m3.842s
sys	0m0.745s

dhruv replied:
Please see #21165
